Yafc.Model/Analysis/DependencyNode.cs

```diff
@@ -0,0 +1,296 @@
+    /// <returns><see langword="true"/> if the known-accessible dependent items are adequate to make the owning item accessible,
+    /// or <see langword="false"/> otherwise.</returns>
+    internal abstract bool IsAccessible(Func<FactorioId, bool> isAccessible);
+
+    /// <summary>
+    /// Gets the additional <see cref="Bits"/> that should be added to this tree's owner, based on the bits <paramref name="getBits"/> returns
+    /// for dependent objects, and the types of this node and its children.
+    /// </summary>
+    /// <param name="getBits">A delegate that returns the <see cref="Bits"/> for a given <see cref="FactorioId"/>.</param>
+    /// <returns>The <see cref="Bits"/> for all the dependent items, ORed together appropriately.</returns>
+    internal abstract Bits AggregateBits(Func<FactorioId, Bits> getBits);
+
+    /// <summary>
+    /// Gets the new <see cref="AutomationStatus"/> for this tree's owner, based on the current value and the statuses returned by
+    /// <paramref name="isAutomatable"/>.
+    /// </summary>
+    /// <param name="isAutomatable">A delegate that returns the <see cref="AutomationStatus"/> for a given <see cref="FactorioId"/>.</param>
+    /// <param name="automationState">The initial <see cref="AutomationStatus"/>, which will not be exceeded regardless of the values returned
+    /// by <paramref name="isAutomatable"/>.</param>
+    /// <returns>The highest feasible <paramref name="automationState"/>, as constrained by <paramref name="isAutomatable"/> and
+    /// <paramref name="automationState"/>.</returns>
+    internal abstract AutomationStatus IsAutomatable(Func<FactorioId, AutomationStatus> isAutomatable, AutomationStatus automationState);
+
+    /// <summary>
+    /// Instructs this dependency tree to draw itself on the specified <see cref="ImGui"/>.
+    /// </summary>
+    /// <param name="gui">The drawing destination.</param>
+    /// <param name="builder">A delegate that will draw the passed <see cref="DependencyList"/> onto the passed <see cref="ImGui"/>.</param>
+    public abstract void Draw(ImGui gui, Action<ImGui, DependencyList> builder);
+
+    /// <summary>
+    /// A <see cref="DependencyNode"/> that requires all of its children.
+    /// </summary>
+    private sealed class AndNode : DependencyNode {
+        private readonly DependencyNode[] dependencies;
+
+        private AndNode(DependencyNode[] dependencies) => this.dependencies = dependencies; // Use Create
+
+        /// <summary>
+        /// Returns a <see cref="DependencyNode"/> that requires all of the children specified in <paramref name="dependencies"/>.
+        /// </summary>
+        /// <param name="dependencies">The children that should all be required by the returned value.</param>
+        internal static DependencyNode Create(IEnumerable<DependencyNode> dependencies) {
+            if (dependencies.Count() == 1) {
+                return dependencies.First();
+            }
```

Comment:
I did convince myself this check is unnecessary. Removing it but keeping the other one just means `DependencyNode.RequireAll(/* a single AndNode */)` will unpack the existing `AndNode` and construct a new one, instead of returning the existing one.